resolver/resolver.cc

```diff
@@ -1317,16 +1374,58 @@ class ResolveConstantsWalk {
     }
 
 public:
-    ResolveConstantsWalk() : nesting_(nullptr) {}
+    ResolveConstantsWalk() : nesting_(nullptr), loadScopeDepth_(0) {
+        firstDefinitionLocs = make_unique<UnorderedMap<core::SymbolRef, core::LocOffsets>>();
+    }
 
-    void preTransformClassDef(core::Context ctx, ast::ExpressionPtr &tree) {
-        nesting_ = make_unique<Nesting>(std::move(nesting_), ast::cast_tree_nonnull<ast::ClassDef>(tree).symbol);
+    void preTransformMethodDef(core::Context ctx, ast::ExpressionPtr &tree) {
+        ENFORCE(loadScopeDepth_ >= 0);
+        loadScopeDepth_++;
+    }
+
+    void postTransformMethodDef(core::Context ctx, ast::ExpressionPtr &tree) {
+        ENFORCE(loadScopeDepth_ > 0);
+        loadScopeDepth_--;
+    }
+
+    void preTransformBlock(core::Context ctx, ast::ExpressionPtr &tree) {
+        ENFORCE(loadScopeDepth_ >= 0);
+        loadScopeDepth_++;
+    }
+
+    void postTransformBlock(core::Context ctx, ast::ExpressionPtr &tree) {
+        ENFORCE(loadScopeDepth_ > 0);
+        loadScopeDepth_--;
     }
 
     void postTransformUnresolvedConstantLit(core::Context ctx, ast::ExpressionPtr &tree) {
         walkUnresolvedConstantLit(ctx, tree);
     }
 
+    void preTransformClassDef(core::Context ctx, ast::ExpressionPtr &tree) {
+        auto &original = ast::cast_tree_nonnull<ast::ClassDef>(tree);
+        auto sym = original.symbol;
+
+        // Populate local first definitions table for out-of-order reference checking.
+        // This happens *iff* the symbol is:
+        //   a). defined only in this file and nowhere else
+        //   b). defined multiple times in this file
+        //   c). this particular classDef is the first occurrence of the symbol in the file
+        if (!ctx.file.data(ctx).isRBI() && loadScopeDepth_ == 0 && sym.isOnlyDefinedInFile(ctx.state, ctx.file)) {
+            auto defLoc = sym.data(ctx)->loc().offsets();
+            auto declLoc = original.declLoc;
+
+            if (defLoc.beginPos() >= declLoc.endPos() && firstDefinitionLocs->find(sym) == firstDefinitionLocs->end()) {
+                // When this condition is met, effectively it means the file has multiple definitions of the symbol
```

Comment:
We should really add some documentation on how `loc` and `declLoc` in `ClassDef` differ, because I always forgot, and as a result, I can't convince myself that this doesn't fire for things that have just single definitions in one file.  I don't think that matters (we still want out-of-order checking in that case?), but just thinking out loud.

Reply:
Will add this comment:

We want `declLoc` here because we want the specific loc of the current def, not the canonical `loc` of the symbol. In case of a single definition, the values will be the same and the check won't fire.